src/StageStridedLoads.cpp

```diff
@@ -0,0 +1,366 @@
+#include "StageStridedLoads.h"
+#include "CSE.h"
+#include "IREquality.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "IRVisitor.h"
+#include "Scope.h"
+#include "Simplify.h"
+#include "Substitute.h"
+#include "UniquifyVariableNames.h"
+
+namespace Halide {
+namespace Internal {
+
+namespace {
+
+class FindStridedLoads : public IRVisitor {
+public:
+    struct Key {
+        // The buffer being accessed.
+        std::string buf;
+        // The base index being accessed, without any constant offset.
+        Expr base;
+
+        // The stride and lanes of the vector access.
+        int64_t stride;
+        int lanes;
+
+        // The loaded type.
+        Type type;
+
+        // The Allocate node the load belongs to. nullptr for loads from external buffers.
+        const Allocate *allocation;
+
+        // The Stmt over which the load definitely happens, and definitely
+        // refers to the same buffer as other loads with the same name. nullptr
+        // means global scope.
+        const IRNode *scope;
+
+        bool operator<(const Key &other) const {
+            // Check fields in order of cost to compare
+            if (stride < other.stride) {
+                return true;
+            } else if (stride > other.stride) {
+                return false;
+            } else if (lanes < other.lanes) {
+                return true;
+            } else if (lanes > other.lanes) {
+                return false;
+            } else if (scope < other.scope) {
+                return true;
+            } else if (scope > other.scope) {
+                return false;
+            } else if (allocation < other.allocation) {
+                return true;
+            } else if (allocation > other.allocation) {
+                return false;
+            } else if (type < other.type) {
+                return true;
+            } else if (other.type < type) {
+                return false;
+            } else if (buf < other.buf) {
+                return true;
+            } else if (buf > other.buf) {
+                return false;
+            } else {
+                return graph_less_than(base, other.base);
+            }
+        }
+    };
+    // Entry entry maps from an offset from the base to a vector of identical
+    // Load nodes with that offset.
+    std::map<Key, std::map<int64_t, std::vector<const Load *>>> found_loads;
+
+    // The current scope over which accesses definitely occur.
+    const IRNode *scope = nullptr;
+
+    Scope<const Allocate *> allocation_scope;
+
+    std::map<const IRNode *, const IRNode *> parent_scope;
+
+protected:
+    void visit(const Load *op) override {
+        if (is_const_one(op->predicate)) {
+            Expr idx = substitute_in_all_lets(simplify(common_subexpression_elimination(op->index)));
```

Comment:
Why is this series of transformations necessary? Hasn't simplification already been run before this pass?

Reply:
A comment might help here